Add Inner blocks to order summary

mikejolley opened the pull request:
The WIP prefix is a leftover, the work on this branch was completed.

This PR converts the Cart Block and Checkout Block "order summary" areas into an inner block structure. 

Cart Block:

- Heading `woocommerce/cart-order-summary-heading`
- Subtotal `woocommerce/cart-order-summary-subtotal`
- Fee `woocommerce/cart-order-summary-fee`
- Shipping `woocommerce/cart-order-summary-shipping`
- Discount `woocommerce/cart-order-summary-discount`
- Coupon Form `woocommerce/cart-order-summary-coupon-form`
- Taxes `woocommerce/cart-order-summary-taxes`

Checkout Block

- Cart Items `woocommerce/checkout-order-summary-cart-items`
- Subtotal `woocommerce/checkout-order-summary-subtotal`
- Fee `woocommerce/checkout-order-summary-fee`
- Shipping `woocommerce/checkout-order-summary-shipping`
- Discount `woocommerce/checkout-order-summary-discount`
- Coupon Form `woocommerce/checkout-order-summary-coupon-form`
- Taxes `woocommerce/cart-order-summary-taxes`

**NOTE:** Yes there are many similarities between the cart and checkout inner blocks, however, they cannot be shared due to the way we use context providers. If they were shared, both Cart and Checkout Blocks would import the above inner blocks, and those inner blocks would then receive the wrong values from context depending on what was imported first. To work around this we have a unique set of inner blocks per each parent block.

The new inner blocks implement existing slotfills, making this change backwards compatible. Subscriptions extension can be installed to verify this since it extends the order totals area including its own shipping and total row.

Fixes #6059

### Screenshots

#### Cart block before

**The Order summary block**

<img width="747" alt="Screenshot 2022-03-29 at 15 47 30" src="https://user-images.githubusercontent.com/1628454/160852502-888753fe-eb6a-4cba-81af-e8e12d042e71.png">

#### Cart block after

**The Order summary block**
<img width="610" alt="Screenshot 2022-03-29 at 15 48 20" src="https://user-images.githubusercontent.com/1628454/160852500-51337fac-59ad-48f7-8bf0-642c75c61c7f.png">

**Selection of an Order summary Inner block**
<img width="611" alt="Screenshot 2022-03-29 at 15 48 32" src="https://user-images.githubusercontent.com/1628454/160852488-30f8277c-ebd3-4edc-86a3-00bcd451b71d.png">





<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

### Testing

#### Automated Tests

Existing block tests cover these inner blocks since the functionality has just moved down a level.

* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [x] E2E tests
  
### Manual Testing

How to test the changes in this Pull Request:

1. Go to a page and add the Checkout block.
2. Go to a page and add the Cart block
3. In the editor, inside each block, select the Order summary and notice that you can select the following inner blocks:

| Cart | Checkout |
| --------------- | --------------- | 
|  Heading, Subtotal, Cart Items, Fees, Coupon Form, Shipping, Taxes | Subtotal, Cart Items, Fees, Coupon Form, Shipping, Taxes | 
4. Try to move some of the inner blocks and save. Confirm that in the Cart block only Order Summary Heading and the Coupon form inner blocks can be removed, and for the Checkout block only the Coupon form.
5. Check on the website that your changes are reflected. Note that Taxes, Fees and Coupons will appear on the website only if the shop supports them.
6. For the Cart > Shipping inner block make sure you can enable disable the shipping calculator and that the change is visibile in the website
7. Enable Taxes on your website (`/wp-admin/admin.php?page=wc-settings`) by clicking  `Enable tax rates and calculations`
8. Go to the C & C blocks and select the Taxes inner block. 
9. In the inner block's setting toggle on / of `Show rate after tax name` and make sure that is reflected in the website
10. Try to place an order and make sure it's successful.




### User Facing Testing

These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).
* [x] Same as above, or
* [ ] See steps below.


### Changelog

> The order summary area for the Cart and Checkout Blocks is now powered by Inner Blocks allowing for more customizations and extensibility.

mikejolley committed:
Sub/Total/Fee inner blocks

mikejolley committed:
Row blocks within the inner block

mikejolley committed:
Update icons

mikejolley committed:
Resolve stying issues

mikejolley committed:
Remove old block

mikejolley committed:
Pin totals row

mikejolley committed:
Locking logic update

mikejolley committed:
Heading inner block

mikejolley committed:
Refactor where inner blocks are defined

mikejolley committed:
Add todos

mikejolley committed:
Todo for Consider deprecating OrderMetaSlotFill and DiscountSlotFill in favour of inner block areas.

mikejolley committed:
Improve frontend registration of components using new entrypoint

mikejolley committed:
Experiment- external block context

mikejolley committed:
Revert "Experiment- external block context"

This reverts commit 4b75668ec7eb62f065c6a488cd942a666e26204f.

mikejolley committed:
Duplicate inner blocks to avoid conflicts with context

mikejolley committed:
Remove todo

mikejolley committed:
Rename block dir

mikejolley committed:
Some test fixes

mikejolley committed:
Fix import

mikejolley committed:
fix import

mikejolley committed:
linting

mikejolley committed:
Remove unused attributes

mikejolley committed:
Optional classname

senadir committed:
fix coupons import

senadir committed:
fix shipping mocks

mikejolley committed:
Styling

mikejolley committed:
Fix selectors in e2e tests

ralucaStan commented on assets/js/blocks/cart/inner-blocks/index.tsx:
```diff
@@ -5,8 +5,15 @@ import './filled-cart-block';
+import './cart-order-summary-shipping';
+import './cart-order-summary-coupon-form';
+import './cart-order-summary-taxes';
+import './cart-order-summary-heading';
```
Minor, but for consistency reasons these should also have `-block` suffix. `./cart-order-summary-heading-block` instead of `./cart-order-summary-heading`

ralucaStan commented on assets/js/blocks/cart/editor.scss:
```diff
@@ -29,3 +29,15 @@ body.wc-lock-selected-block--remove {
+	.wc-block-components-totals-coupon,
+	.wc-block-components-order-summary,
+	.wc-block-components-totals-shipping {
+		box-sizing: border-box;
```
Before we had the `.wc-block-components-totals-wrapper` class for each item of the summary.
Now:
- we use the `.wc-block-components-totals-wrapper` class only for slots
- we have `.wc-block-components-totals-item` for some of the items 
- summary, coupon and shipping don't have the `.wc-block-components-totals-item` and it needs to be styled separately.

I think it was good to have one class for each of this summary items. I wonder why:

- we didn't keep the `.wc-block-components-totals-wrapper`? 
- we don't add `.wc-block-components-totals-item` to coupon, summary and shipping as well?

ralucaStan commented on assets/js/base/components/cart-checkout/totals/footer-item/index.tsx:
```diff
@@ -30,6 +31,8 @@ export interface TotalsFooterItemProps {
 	 * convenience, but will use only these two properties.
 	 */
 	values: LooselyMustHave< CartResponseTotals, 'total_price' | 'total_tax' >;
+	// CSS classname.
```
let's remove this

ralucaStan commented on assets/js/blocks/cart/inner-blocks/cart-order-summary-coupon-form/block.tsx:
```diff
@@ -0,0 +1,26 @@
+
+	return (
+		<TotalsCoupon
+			className={ className }
```
The `Additional CSS class(es)` attribute doesn't work for this block.

ralucaStan replied:
I also wonder why we don't show the `.wp-block-woocommerce-cart-order-summary-coupon-form-block` class like for the other inner blocks `.wp-block-woocommerce-cart-order-summary-NAME-block`.

ralucaStan replied:
both fixed

ralucaStan commented on assets/js/blocks/cart/inner-blocks/cart-order-summary-heading/attributes.ts:
```diff
@@ -0,0 +1,18 @@
+		type: 'object',
+		default: {
+			remove: false,
+			move: false,
```
Should we allow to move/remove this considering the corresponding title in the left side is not an inner block? Maybe we can enable this once that becomes an inner title as well?
<img width="670" alt="Screenshot 2022-03-29 at 19 12 12" src="https://user-images.githubusercontent.com/1628454/160667923-0140d95e-06ed-41e5-acac-309168b1a0f2.png">
I

ralucaStan replied:
After internal discussions it was decided that  we can leave it since it’ll be up to the user to remove it. And he can add a custom CSS to add some margin to fix the alignment issue.

ralucaStan commented on assets/js/blocks/cart/inner-blocks/register-components.ts:
```diff
@@ -0,0 +1,185 @@
+} );
+
+registerCheckoutBlock( {
+	metadata: metadata.FILLED_CART,
```
this is registered a second time, it should be removed.

ralucaStan reviewed (commented):
Posting a partial review comment for the Cart Inner blocks, I'll still need to check the Checkout inner blocks

ralucaStan commented on assets/js/blocks/checkout/inner-blocks/index.tsx:
```diff
@@ -13,3 +13,10 @@ import './checkout-order-summary-block';
+import './checkout-order-summary-shipping';
+import './checkout-order-summary-coupon-form';
+import './checkout-order-summary-taxes';
+import './checkout-order-summary-cart-items';
```
these should also end with `-block`

ralucaStan commented on assets/js/blocks/checkout/styles/editor.scss:
```diff
@@ -72,3 +72,15 @@ body.wc-lock-selected-block--remove {
+	.wc-block-components-order-summary,
+	.wc-block-components-totals-shipping {
+		box-sizing: border-box;
+	}
```
same comment about adding one common class for all section wrappers from the Order summary column

ralucaStan commented on assets/js/blocks/checkout/inner-blocks/checkout-order-summary-coupon-form/block.tsx:
```diff
@@ -0,0 +1,30 @@
+
+	return (
+		<TotalsCoupon
+			className={ className }
```
The coupon form wrapper does not apply the additional class provided by the user.

ralucaStan reviewed (changes requested):
I've also added my comments about the Checkout inner blocks.

All and all this works as expected. There are some layout shifts when selecting the inner blocks, but I will create a separate issue about this.

I've added some comments that need to be addressed.

ralucaStan commented:
It was a real effort reviewing this as I imagine was the case for writing this. 
It would have helped with testing if we would have had separate PRs for Cart and Checkout innerblocks. In this case the decisions made for the first block would have been applied to the 2nd block.

ralucaStan committed:
Add back the wc-block-components-totals-wrapper class that was used for each segment in the totals Order summary

Because, removing them was:
-  a breaking change for the old structure
- was making it harder to target the inner blocks. Before the class was used to target each segment
- it was making the wc-block-components-totals-item behave as a child or parent depending on the inner block, inconsitency

ralucaStan committed:
Reuse the TotalsWrapper component for C& C blocks inner blocks

This component was removed in this PR, but  we wrap components in the Cart and Checkout sidebar in a TotalsWrapper. This will ensure consistent spacing and borders are applied to items in the sidebar.

ralucaStan commented:
I've added two commits to address the changes requests [10670393a00ebd18d12b86755b5278cdfa49ace6](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6065/commits/10670393a00ebd18d12b86755b5278cdfa49ace6) and https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6065/commits/d654f5b074932b68a218ae5fc319d1b4444133a4:

### Add back the wc-block-components-totals-wrapper class that was used for each segment in the totals Order summary

Because, removing them was:
-  a breaking change for the old structure
- was making it harder to target the inner blocks. Before the class was used to target each segment
- it was making the wc-block-components-totals-item behave as a child or parent depending on the inner block, inconsitency

### Reuse the TotalsWrapper component for C& C blocks inner blocks
This component was removed in this PR, but  we wrap components in the Cart and Checkout sidebar in a TotalsWrapper. This will ensure consistent spacing and borders are applied to items in the sidebar.
